src/WallpaperEngine/Render/CVideo.cpp

```diff
@@ -4,31 +4,92 @@ using namespace WallpaperEngine;
 
 using namespace WallpaperEngine::Render;
 
-CVideo::CVideo (Core::CVideo* video, irr::video::IVideoDriver* driver) :
-    CWallpaper (video, Type),
-    m_driver (driver)
+CVideo::CVideo (Core::CVideo* video, WallpaperEngine::Irrlicht::CContext* context) :
+    CWallpaper (video, Type, context)
 {
+}
+
+void CVideo::render ()
+{
+    irr::video::IVideoDriver* driver = m_context->getDevice ()->getVideoDriver ();
     int width = driver->getScreenSize ().Width;
     int height = driver->getScreenSize ().Height;
-    video->initFrames (width, height);
-    m_frameImage = m_driver->createImage (irr::video::ECOLOR_FORMAT::ECF_R8G8B8,
+
+    m_frameImage = m_context->getDevice ()->getVideoDriver ()->createImage (irr::video::ECOLOR_FORMAT::ECF_R8G8B8,
                 irr::core::dimension2du(width, height));
+
+    getVideo ()->setSize (width, height);
+    getNextFrame ();
+    writeFrameToImage ();
+
+    driver->removeTexture (m_frameTexture);
```

Comment:
It should be possible to reuse the same texture and image so we don't need to allocate a new one for every frame being rendered (this would be better performance-wise), but looking at the documentation it isn't that clear if locking the texture allows us to use the buffer returned to write data directly as it comes from the video (based on some testing I've done, i'd say that the image format isn't the same, but I'm not 100% certain on that). I'll create a task for this after merging so someone can look into it.

Reply:
I tried doing that but I was not able to find a way because of the rescaling. Is there a way to change the width and height of `irr::video::IImage` after it was initialized?

Reply:
I don't think so, resizing the textures could lead to allocations anyway.

I'd say the best way to approach this is to expose the "m_isRootWindow" and "m_viewports" in CContext first so on initialization we can create the required textures in case of being displayed as desktop background (the app doesn't support changing screen settings yet) and save them under a std::vector, then the render function can cycle through the textures using an iterator. For window mode we would need to check the viewport's size from the driver and destroy the old texture and create a new one if the size changes.

As I said, we can look into this later, It's just optimization, what's more important so we can get this merged is this comment (I don't know why github didn't hightlight it btw): 
https://github.com/Almamu/linux-wallpaperengine/pull/14/files/ab937780ed2d1e55bff4c53f5c287b3a2dca23aa..3079b42cc4a7a17c0d8bd3afdd462d6363cc4ee7#r408581373